### PR TITLE
Add mirror gateway definitions and Implementations

### PR DESCRIPTION
**Pull request by hlts2**
#### WHAT

I have added a new component(`Vald Mirror Gateway`) for BCP. 
Vald Mirror Gateway is an optional component of the Vald, allowing the vector data to be synchronized across multiple Vald clusters. This component makes it possible to enhance availability during a cluster failure.
The following RPCs will be added in this PR. The CRUD RPCs remain unchanged.
- Register RPC: Register mirror servers.

#### WHY

we need to support request mirroring for BCP

#### :warning: FYI

Although there are many differences, most of them are due to automatic generation. 
The following is the main addition.

- https://github.com/vdaas/vald/blob/feature/mirror-gateway-definition/apis/proto/v1/mirror/mirror.proto
- https://github.com/vdaas/vald/blob/feature/mirror-gateway-definition/apis/proto/v1/payload/payload.proto#L525

### Related PR:

https://github.com/vdaas/vald/pull/1792
https://github.com/vdaas/vald/pull/1948
https://github.com/vdaas/vald/pull/2237

### Versions:

- Go Version: 1.19.5
- Docker Version: 20.10.8
- Kubernetes Version: 1.22.0
- NGT Version: 2.0.9

### Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/main/CONTRIBUTING.md) document and completed [our CLA agreement](https://cla-assistant.io/vdaas/vald).
- [x] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?

#### How to deploy

1. Update helm repo for monitoring 
```
helm repo update
```

2. Create k3d cluster
```
make k3d/restart
```

3. Deploy monitoring stack
```
 make k8s/external/cert-manager/deploy k8s/monitoring/deploy
```

4. Deploy multiple vald clusters
```
 make k8s/multi/vald/deploy
```

4.1. The following resources(`MirrorTarget`) is deployed using above command.

```
---
apiVersion: vald.vdaas.org/v1
kind: ValdMirrorTarget
metadata:
  name: mirror-target-01
  labels:
    group: mirror-group-01
spec:
  colocation: dc1
  target:
    host: vald-mirror-gateway.vald-01.svc.cluster.local
    port: 8081
---
apiVersion: vald.vdaas.org/v1
kind: ValdMirrorTarget
metadata:
  name: mirror-target-02
  labels:
    group: mirror-group-01
spec:
  colocation: dc1
  target:
    host: vald-mirror-gateway.vald-02.svc.cluster.local
    port: 8081
```

6. (E2E test)

Please execute the following command if you want to perform the e2e test.
```
E2E_TARGET_NAME=vald-mirror-gateway E2E_TARGET_NAMESPACE=vald-03 make e2e
```

### CODE REVIEW

File: charts/vald/values/multi-vald/dev-vald-with-mirror.yaml [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+defaults:
+  image:
+    tag: pr-2237 # TODO: Delete it later.
```
Comment (hlts2) [inline charts/vald/values/multi-vald/dev-vald-with-mirror.yaml:18]:
```suggestion
    tag: pr-2262 # TODO: Delete it later.
```

File: docs/user-guides/mirroring-configuration.md [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    enabled: true
+```
+
+If you want to make more detailed settings, please set the following parameters.
+
+```yaml
+gateway:
+  mirror:
+  ...
+    gateway_config:
```
[the remark this answers is not in the document]
Comment (hlts2) [reply docs/user-guides/mirroring-configuration.md:40]: > **NOTE**
It is a configuration value. So we do not need to fix it

File: docs/user-guides/mirroring-configuration.md [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    ...
+      # gRPC client configuration (overrides defaults.grpc.client)
+      client: {}
+      # The duration to register other Mirror Gateways.
+      register_duration: "1s"
```
[the remark this answers is not in the document]
Comment (hlts2) [reply docs/user-guides/mirroring-configuration.md:45]: same as the following comment.
https://github.com/vdaas/vald/pull/2262#discussion_r1418376233

File: docs/user-guides/mirroring-configuration.md [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+      # The target namespace to discover ValdMirrorTarget (CR) resource.
+      # The default value is its own namespace.
+      namespace: "vald"
+      # The group name of the Mirror Gateways (optional).
+      # It is used to discover ValdMirrorTarget resources (CR) with the same group name.
+      # The default value is empty.
+      group: "group1"
```
[the remark this answers is not in the document]
Comment (hlts2) [reply docs/user-guides/mirroring-configuration.md:52]: > **NOTE**
Same as the following comment.
https://github.com/vdaas/vald/pull/2262#discussion_r1418376233

File: docs/user-guides/mirroring-configuration.md [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+
+```yaml
+apiVersion: vald.vdaas.org/v1
+kind: ValdMirrorTarget
+metadata:
+  name: mirror-target-01
+  namespace: vald-03
+  labels:
+    # The group name of the Mirror Gateways.
+    group: mirror-group-01
```
[the remark this answers is not in the document]
Comment (hlts2) [reply docs/user-guides/mirroring-configuration.md:79]: > **NOTE**
Same as the following comment.
https://github.com/vdaas/vald/pull/2262#discussion_r1418376233

File: docs/user-guides/mirroring-configuration.md [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+      register_duration: "1s"
+      # The target namespace to discover ValdMirrorTarget (CR) resource.
+      # The default value is its own namespace.
+      namespace: "vald"
+      # The group name of the Mirror Gateways (optional).
+      # It is used to discover ValdMirrorTarget resources (CR) with the same group name.
+      # The default value is empty.
+      group: "group1"
+      # The duration to discover other mirror gateways in the same group.
+      discovery_duration: 1s
```
[the remark this answers is not in the document]
Comment (hlts2) [reply docs/user-guides/mirroring-configuration.md:54]: > **NOTE**
Same as the following comment.
https://github.com/vdaas/vald/pull/2262#discussion_r1418376233

File: docs/user-guides/mirroring-configuration.md [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+  ...
+    gateway_config:
+    ...
+      # gRPC client configuration (overrides defaults.grpc.client)
+      client: {}
+      # The duration to register other Mirror Gateways.
+      register_duration: "1s"
+      # The target namespace to discover ValdMirrorTarget (CR) resource.
+      # The default value is its own namespace.
+      namespace: "vald"
```
[the remark this answers is not in the document]
Comment (hlts2) [reply docs/user-guides/mirroring-configuration.md:48]: > **NOTE**
Same as the following comment.
https://github.com/vdaas/vald/pull/2262#discussion_r1418376233

**Review by datelier — APPROVED** [review]

File: internal/k8s/vald/mirror/target/target.go [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+		if apierr.IsNotFound(err) {
+			log.Errorf("not found: %s", err)
+			return reconcile.Result{
+				Requeue:      true,
+				RequeueAfter: time.Second,
+			}, nil
+		}
+		return reconcile.Result{
+			Requeue:      true,
+			RequeueAfter: 100 * time.Millisecond,
```
[the remark this answers is not in the document]
Comment (hlts2) [reply internal/k8s/vald/mirror/target/target.go]: Thank you for your comment 🙏 
I have fixed it based on your comment!
https://github.com/vdaas/vald/pull/2262/commits/62326006d0abc37fadfc8ee06aae98804226880f

File: internal/k8s/vald/mirror/target/target.go [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+	r.lopts = append(r.lopts, opt)
+}
+
+func (r *reconciler) Reconcile(ctx context.Context, _ reconcile.Request) (res reconcile.Result, err error) {
+	ml := &mirrv1.ValdMirrorTargetList{}
+	err = r.mgr.GetClient().List(ctx, ml, r.lopts...)
+	if err != nil {
+		r.onError(err)
+		if apierr.IsNotFound(err) {
+			log.Errorf("not found: %s", err)
```
[the remark this answers is not in the document]
Comment (hlts2) [reply internal/k8s/vald/mirror/target/target.go]: Thank you for your comment 🙏
In the first place, the log output is not needed in this case, so I removed the code.
https://github.com/vdaas/vald/commit/62326006d0abc37fadfc8ee06aae98804226880f

File: internal/k8s/reconciler.go [hunk 10]
```diff
@@ -25,6 +25,7 @@ import (
 	"github.com/vdaas/vald/internal/net"
 	"github.com/vdaas/vald/internal/safety"
 	"github.com/vdaas/vald/internal/sync/errgroup"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
```
[the remark this answers is not in the document]
Comment (hlts2) [reply internal/k8s/reconciler.go]: Ah, thanks for catching it!!
I have fixed it with this commit 👍 https://github.com/vdaas/vald/pull/2262/commits/62326006d0abc37fadfc8ee06aae98804226880f

File: internal/k8s/reconciler.go [hunk 11]
```diff
@@ -34,10 +35,15 @@ import (
...
 )
 
+type Manager = manager.Manager
+
 type Controller interface {
 	Start(ctx context.Context) (<-chan error, error)
+	GetManager() Manager
 }
 
+var Now = v1.Now
```
[the remark this answers is not in the document]
Comment (hlts2) [reply internal/k8s/reconciler.go]: Ah, thanks for catching it!!
I have fixed it with this commit 👍 https://github.com/vdaas/vald/commit/62326006d0abc37fadfc8ee06aae98804226880f

**Review by vankichi — DISMISSED** [review]

**Review by ykadowak — APPROVED** [review]

**Review by ykadowak — DISMISSED** [review]

File: pkg/gateway/mirror/handler/grpc/handler.go [hunk 12]
```diff
@@ -0,0 +1,3344 @@
...
+	vAddr             string // Vald gateway address (LB gateway).
+	streamConcurrency int
+	name              string
+	ip                string
+	vald.UnimplementedValdServerWithMirror
+}
+
+const apiName = "vald/gateway/mirror"
+
+// New returns a Vald server as as gRPC handler with mirror using the provided options.
```
[the remark this answers is not in the document]
Comment (hlts2) [reply pkg/gateway/mirror/handler/grpc/handler.go:51]:
```suggestion
// New returns a Vald server as gRPC handler with mirror using the provided options.
```

**Review by vankichi — APPROVED** [review]
LGTM

**Review by ykadowak — APPROVED** [review]

**Comment by hlts2** [comment]
@vankichi @ykadowak 
Thank you for your previous review 🙏 
Sorry, I updated the license header and the approval disappeared....
There is no change to the logic.

**Review by ykadowak — DISMISSED** [review]

**Review by datelier — APPROVED** [review]

**Review by kpango — APPROVED** [review]
LGTM

**Review by vankichi — DISMISSED** [review]
LGTM

**Review by ykadowak — APPROVED** [review]

**Review by kpango — APPROVED** [review]
LGTM

**Review by vankichi — APPROVED** [review]

**Review by ykadowak — APPROVED** [review]

**Review by kpango — APPROVED** [review]

**Review by datelier — APPROVED** [review]

**Comment by hlts2** [comment]
Thank you for the review!! 🙏 
I will merge this PR!!!

**Pull request merged**